### PR TITLE
Added support for Size Classes.

### PR DESCRIPTION
**Pull request by DandyLyons**
See: https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/adaptivity-and-layout/ (at Size Classes)

### CODE REVIEW

**Comment by DandyLyons** [comment]
This is my very first pull request. 
Hopefully this is helpful.

**Comment by zntfdr** [comment]
Hi @DandyLyons! Thank you so much for this 😃

This is an awesome addition, however as it is it breaks the current behavior 😅 

I believe we could add a new [option set](https://developer.apple.com/documentation/swift/optionset) parameter where we can declare which change the stack reacts to.

For example: 
- `AHStack(observing: [.sizeCategory, .sizeClass])` would change orientation both when the (horizontal) size class is compact and/or when the size category is among the accessibility ones.
- `AHStack(observing: [.sizeClass])` would change orientation only when the (horizontal) size class is compact.
- `AHStack(observing: [.sizeCategory])` would change orientation only when the size category is among the accessibility ones (current behavior).
- `AHStack(observing: [])` would never change 😆.

**Comment by DandyLyons** [comment]
Wow, that's a great idea. I haven't implemented option sets before but I think I've seen that in other API's. 

The bummer of that approach is it changes the call site. I was so spoiled before. I loved being able to simply add A to an HStack and now all of a sudden it's adaptable.

**Comment by DandyLyons** [comment]
Also, I'd love to eventually add the functionality of your FlexibleView to an AHStack. 

Perhaps with a call site like 'AHStack(observing: [.sizeClass, .sizeCategory, .currentWidth])'

**Comment by zntfdr** [comment]
Agreed, that would complicate things quite a bit 😅  
I'm holding big hopes for #WWDC21 for a native way to do so 🙏🏻

**Comment by DandyLyons** [comment]
I just pushed a new update. Not sure if it's already in the pull request or not. 

I decided to go with `horizontalSizeClass` and `verticalSizeClass` instead of `.sizeClass` so that we could be more consistent with the `@Environment(\.horizontalSizeClass)` that Apple uses.

**Comment by zntfdr** [comment]
Thank you Dandy! Will review and merge it soon! 😊

**Comment by DandyLyons** [comment]
I've discovered a bug in my code. Haven't found the cause yet. 

When AHStack is the root view of the body and it's observing horizontalSizeClass, it behaves as if it's in .regular, even if it's actually in .compact

**Comment by DandyLyons** [comment]
Actually it appears to be just an issue with the Canvas preview. When I run it in the simulator, I don't see this issue. I'll keep testing.

**Comment by zntfdr** [comment]
Thank you again, Dandy! 🚀 

FYI before release I will make it that the default behavior is as per current release, so people that update to this version won't see a change all of a sudden 😊

**Comment by DandyLyons** [comment]
Love it! Thank you for everything. 

(Haha, I meant to match the default behavior to the current release, but I guess I got confused and did the opposite by accident. 😅 Thanks for catching that.) 

Also, good catch on making it compatible with macOS, tvOS, and watchOS. I didn't think about that.

**Comment by DandyLyons** [comment]
I don't know if this behavior would affect AStack at all but I was playing around with Apple's Fruta app example and discovered this bug (feature?). They have a similar approach that switches between 2 layout configurations depending on size class (but I think AStack feels a little more elegant 😁). Anyways, I noticed that when I dynamically switch size classes in Fruta, it forgets the selection state. (Example screen recording attached here: https://drive.google.com/file/d/1GUhToU8btQ5HuoWLUPvYWOlZW3hC2ymn/view?usp=sharing)

Anyways, this bug probably has nothing to do with AStack. (My guess is it might be a bug of NavigationView?). But I thought it might interest you for one of your blog posts.

**Comment by zntfdr** [comment]
Thank you for letting me know! 
It most likely depends on where the `@State` is stored: if the state is stored within the layout that is being changed, it will be reset.

I believe it's the same in `AStack`: when changing layout we're destroying the current content and re-create it from scratch with the new layout. We don't and can't move the current content from one stack to the other. 

I've filed a feedback on this, please feel free to dupe it :) FB9061954

In short what we would need is a native SwiftUI `Stack` view where the axis is a parameter that we can change, currently this is not possible and we're swapping the entire `VStack`/`HStack` view instead.

PS  
The solution is to store the state elsewhere (out of the `AStack` content), so it won't be reset when changing layout.

**Comment by DandyLyons** [comment]
Perhaps that should be added as a best practices in the README.

**Comment by DandyLyons** [comment]
Hey Federico! Happy Dubdub! 

Unfortunately I don't think Apple created an automatic Adaptive Stack. 😭
But, watching "Demistify SwiftUI" https://developer.apple.com/videos/play/wwdc2021/10022/ really helped me better understand what we were talking about earlier, how SwiftUI forgets state as soon as the AStack adapts. 
Spoke to an Apple engineer regarding this problem which I'll call "the identity problem", and he pretty much confirmed exactly what you were saying: it depends on where `@State` is stored. (Conversation is here: https://wwdc21.slack.com/archives/C023DHQPY5B/p1623277979024700)

So I was wondering if there was a way that we could build this behavior into AStack itself. Perhaps `AHStack` could be the one to store `@State` and calculate `shouldAdapt` whereas `_AHStack` could have the actual `if` statement in it. I think this would solve the identity problem right? Here's just a quick proof of concept. What do you think? 

```swift
import SwiftUI

struct CallSite: View {
    
    var body: some View {
        AHStack {
            Text("Hello")
            Text("World")
        }

    }
}

struct AHStack<Content: View>: View {
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass {
        didSet {
            calculateIfShouldAdapt()
        }
    }
    @State var shouldAdapt: Bool = false
    @ViewBuilder var content: () -> Content
    
    init(@ViewBuilder content: @escaping () -> Content) {
        self.content = content
        
        calculateIfShouldAdapt()
    }
    
    var body: some View {
        _AHStack(shouldAdapt: shouldAdapt) {
            content()
        }
    }
    
    private func calculateIfShouldAdapt() {
        if horizontalSizeClass == .compact {
            shouldAdapt = true
        } else {
            shouldAdapt = false
        }
    }
}

struct _AHStack<Content: View>: View {
    var shouldAdapt: Bool
    @ViewBuilder var content: () -> Content
    
    var body: some View {
        if shouldAdapt {
            VStack {
                content()
            }
        } else {
            HStack {
                content()
            }
        }
    }
}
```

PS 
lol. I can't believe that I accidentally gave @Majid the credit. My bad. I love both of you guys so much.

**Comment by zntfdr** [comment]
Hi Daniel,
thank you for the follow-up 😊

The challenge is that the state **within** `Content` is reset when we move between the two stacks, not the `AStack` state.

E.g. if we declare this view:

```swift
struct MyView: View {
  @State var isOn: Bool = false

  var body: some View {
    Toggle(isOn: $isOn)
  }
}
```

and put it within `AStack`:

```swift
AStack {
  MyView()
}
```

Whenever we switch axes, `MyView`'s `isOn` state is reset to the default value (`= false` in the example above). 

As the engineer suggests, and as I did in the Post Scriptum here https://github.com/zntfdr/AStack/pull/3#issuecomment-810737085, the solution is to store the `Content` state outside `Content`.

Continuing from the example above, this means updating `MyView` to accept a `@Binding`:

```swift
struct MyView: View {
  @Binding var isOn: Bool

  var body: some View {
    Toggle(isOn: $isOn)
  }
}
```

and then store the state somewhere else in a parent, e.g.:

```swift
struct ContentView: View {
  @State var isOn: Bool

  var body: some View {
    AStack {
      MyView(isOn: $isOn)
    }
  }
}
```
 
With this change, switching axis won't reset `MyView` state, because `MyView` doesn't have any, it's `ContentView` that stores the state and pass it down to `MyView`.

Hope this clears things up!

PS

I wrote all the code above off the top of my head, I think I'm missing a `label` parameter in the `Toggle` 😄

**Comment by zntfdr** [comment]
@DandyLyons I've made a small prototype where the state is persisted between changes of branches.

In order to do so I had to create my own property wrapper that restores the state when we switch branches:

https://user-images.githubusercontent.com/5277837/121767929-9c66be00-cb85-11eb-8e77-28a8456f291e.mov

I've also added `matchedGeometryEffect` just for fun, here's the full gist:

```swift
import SwiftUI

var storage: [String: Any] = [:]

@propertyWrapper
struct FSState<V>: DynamicProperty {
  @State var value: V
  var identifier: String

  var wrappedValue: V {
    get { value }
    nonmutating set {
      value = newValue
    }
  }

  var projectedValue: Binding<V> {
    $value
  }

  func update() {
    storage[identifier] = value
  }

  init(identifier: String, defaultValue: V) {
    self.identifier = identifier
    self._value = State(initialValue: (storage[identifier] as? V) ?? defaultValue)
  }
}

struct ContentView: View {
  @State var isTrue = true
  @Namespace var namespace

  var body: some View {
    VStack {
      if isTrue {
        HStack {
          MyView(identifier: "a")
            .matchedGeometryEffect(id: "a", in: namespace)
          MyView(identifier: "b")
            .matchedGeometryEffect(id: "b", in: namespace)
        }
      } else {
        VStack {
          MyView(identifier: "a")
            .matchedGeometryEffect(id: "a", in: namespace)
          MyView(identifier: "b")
            .matchedGeometryEffect(id: "b", in: namespace)
        }
      }
      Button("toggle branch") {
        withAnimation {
          isTrue.toggle()
        }
      }
    }
    .padding()
  }
}

struct MyView: View {
  @FSState var isOn: Bool
  var identifier: String

  init(identifier: String) {
    self.identifier = identifier
    _isOn = .init(identifier: identifier, defaultValue: false)
  }

  var body: some View {
    Toggle(identifier, isOn: $isOn)
  }
}
```

Unfortunately, this won't work with `AStack`:
I guess because the view definition doesn't match exactly what you see above.

With the code in the gist, `@FSState` is re-initialized when we switch axis, therefore allowing the state to be restored.

Unfortunately, if we use `@FSState` with `AStack`, that state is not re-initialized when we switch axis, at least not in the current version of `AStack`, therefore the logic restoring the state in the init is not run, thus losing the state.

Again, the best way to solve this is what I recommended in the previous comment, please consider this an experiment just to learn more about how things work behind the scenes 😊

Thank you for all your questions (here and in the Digital Lounges)!

**Comment by DandyLyons** [comment]
Wow so awesome! I feel like there is definitely a way to do it. 

Just a quick thought, the binding strategy does seem to work reliably as a workaround, but I'm curious what the limitations are. Particularly, I wonder if we'll be able to use this workaround on types that don't accept a binding. If it's a custom type, then I suppose we could just extend it to accept a binding. But what if AStack is wrapping a SwiftUI primitive that doesn't accept a binding. I may be hesitant to extend a primitive to accept a binding because I don't want to mess with the native behavior. 

For that matter I'm not sure if I totally understand how deep into the children tree we need to pass a binding. It looks like we just need to pass it one generation, and then all the grandchildren etc will be protected from invalidation but I just want to make sure. If there were a risk that grandchildren etc. could still be invalidated then that would be a problem, because we'd have to propagate the binding all the way down the children. (If that were the case, maybe we could pass a binding down through the environment.)

**Comment by zntfdr** [comment]
> what the limitations are

The "only" limitation I see is that the state has to be stored in an ancestor.
`@Binding` is just one way:  `@ObservedObject`, `@Environment`, `@EnvironmentObject` are also guaranteed to work, too.

> It looks like we just need to pass it one generation, and then all the grandchildren etc will be protected from invalidation but I just want to make sure.

Yes, as long as the state is stored in a view whose identity is not affected by the `AH/VStack`, it's protected.
